### DISCUSSION TITLE
What is consuming RAM after `on_train_epoch_start` but before `training_step`

### DISCUSSION THREAD

**Discussion by malfonsoarquimea** [Lightning-AI/pytorch-lightning#12522] [Lightning Trainer API: Trainer, LightningModule, LightningDataModule]
Hi! I am currently facing an issue where a lot of ram is consumed while training. By running the debugger while monitoring ram consumption I managed to find that the ram consumption process happens after on_train_epoch_start callback and before training_step, and I don't have any on_train_batch_start callback defined. I think the issue is related to how I defined my dataset and dataloaders, but I can not find what part of the code is producing the memory consumption since I don't know which piece of the code is called in between on_train_epoch_start and training_stepso I can not debug it.
Do you have some insight on where to look to find where is memory being consumed?
Thanks very much in advance

**Comment by rohitgr7** [comment]
it must be the batch that is extracted from your dataloader that is being passed to your `training_step`.

**Reply by malfonsoarquimea** [reply]
Thanks for your answer!
The thing is that the batch itself occupies much less than the memory consumption I experience, here you have some context:

- the ram memory consumption at the end of  on_train_epoch_start() is 5,8 GB
- the ram memory consumption at the beginning of the first training step is 75 GB
- the batch is composed of a tensor of size 1600 with type float32 so it occupies 6400 Bytes or 6.4 KB
- the whole dataset is an uncompressed h5 file with a size of  17GB
- I am using my own custom dataset (a class that inherits from torch.utils.data.Dataset that is passed to the dataloader)
Clearly my dataloader is loading something into memory that I am unable to identify. I cannot track the memory consumption as I don't know where to look, any clue would be appreciated.
thanks in advance!!

**Reply by malfonsoarquimea** [reply]
Hi, @rohitgr7 , I was unable of finding what is being loaded. As I don't know what code is the one responsible of this memory consumption I am not able to profile it. Any suggestion?
Thanks in advance buddy

**Reply by rohitgr7** [reply]
well there is nothing except the batch that is being loaded between these two hooks.

from the codebase, this is the piece of code that gets executed
https://github.com/PyTorchLightning/pytorch-lightning/blob/f300b60f472b56f1aabd428a268184b53d3f9a55/pytorch_lightning/loops/fit_loop.py#L255-L269

https://github.com/PyTorchLightning/pytorch-lightning/blob/f300b60f472b56f1aabd428a268184b53d3f9a55/pytorch_lightning/loops/epoch/training_epoch_loop.py#L151-L208

there is another hook called `on_train_batch_start` that is executed between these hooks. Can you monitor the memory consumption there?

**Reply by malfonsoarquimea** [reply]
Thanks very much @rohitgr7, thanks to your help, i was able to track the memory consumption to line 153 of loops/epoch/training_epoch_loop.py.

https://github.com/PyTorchLightning/pytorch-lightning/blob/f300b60f472b56f1aabd428a268184b53d3f9a55/pytorch_lightning/loops/epoch/training_epoch_loop.py#L153

The code runs flawlessly to that point and there, while executing this line, the RAM consumption starts to increase until all the available ram is consumed. What is exactly done there to produce such a huge memory consumption?

**Reply by rohitgr7** [reply]
is it happening on the latest release? I remember some issues were fixed recently in the data fetcher.

**Reply by malfonsoarquimea** [reply]
@rohitgr7 yes, I am using pytorch lightning 1.6.3

**Reply by rohitgr7** [reply]
ah! then it's an issue in PL. Is it possible for you to share a reproducible script?

**Reply by malfonsoarquimea** [reply]
@rohitgr7 Sorry for the delay, didn't have time to follow this issue. I will try to create a reproducible script so you can debug it!
regards!
